### PR TITLE
config: amend libtirpc detection for opensuse

### LINKED ISSUE #7422: Build failure on OpenSUSE Tumbleweed: tirpc/rpc/xdr.h: No such file or directory
Type                                | Version/Name
  ---                                  |     --- 
Distribution Name       | OpenSUSE
Distribution Version    |  Tumbleweed
Linux Kernel                 | 4.16.0-1-default
Architecture                 | x86-64
ZFS Version                  | git, 74df0c5e251a920a1966a011c16f960cd7ba562e
SPL Version                  | git, 9125f8f5bdb36bfbd2d816d30b6b29b9f89ae3d8

`make rpm` fails at this point:

```
make[5]: Entering directory '/tmp/zfs-build-root-EWGHh2Jo/BUILD/zfs-0.7.0/lib/libnvpair'
  CC       libnvpair.lo
  CC       libnvpair_json.lo
  CC       nvpair_alloc_system.lo
  CC       nvpair_alloc_fixed.lo
  CC       nvpair.lo
  CC       fnvpair.lo
In file included from ../../module/nvpair/nvpair.c:34:0:
../../lib/libspl/include/rpc/xdr.h:41:10: fatal error: tirpc/rpc/xdr.h: No such file or directory
 #include <tirpc/rpc/xdr.h>
          ^~~~~~~~~~~~~~~~~
compilation terminated.
make[5]: *** [Makefile:695: nvpair.lo] Error 1
```

I don't really know what's going on. But for reference, the package libtirpc-devel includes the file `/usr/include/rpc/xdr.h` .

### PR DESCRIPTION
**Pull request by rkitover**
libtirpc-devel on opensuse tumbleweed latest installs headers into
/usr/include/rpc and not /usr/include/tirpc/rpc which is the standard
location.

Try to handle both cases as gracefully as possible.

Preloading the tirpc rpc/xdr.h with -include is necessary to avoid a
conflict with the namesake libspl header.

Tested on tumbleweed which uses the /usr/include/rpc header path and on
gentoo which uses the standard path.

Signed-off-by: Rafael Kitover <rkitover@gmail.com>

### Motivation and Context
ZoL fails to build on opensuse tumbleweed without making a /usr/include/tirpc -> /usr/include symlink.
Fix #7422

### How Has This Been Tested?
Built ZoL for 4.20.0 on opensuse tumbleweed with my changes, this installation is on zfs and works fine.
Built ZoL on gentoo, which uses the standard libtirpc header path.

### Types of changes
- [x] Bug fix (non-breaking change which fixes an issue)

### Checklist:
- [ ] My code follows the ZFS on Linux [code style requirements](https://github.com/zfsonlinux/zfs/blob/master/.github/CONTRIBUTING.md#coding-conventions).
- [ ] I have updated the documentation accordingly.
- [x] I have read the [**contributing** document](https://github.com/zfsonlinux/zfs/blob/master/.github/CONTRIBUTING.md).
- [ ] I have added [tests](https://github.com/zfsonlinux/zfs/tree/master/tests) to cover my changes.
- [ ] All new and existing tests passed.
- [x] All commit messages are properly formatted and contain [`Signed-off-by`](https://github.com/zfsonlinux/zfs/blob/master/.github/CONTRIBUTING.md#signed-off-by).

### CODE REVIEW

**Comment by Conan-Kudo** [comment]
@rkitover This smells like an openSUSE bug rather than something that should be fixed in ZoL. Given that this change exists only in openSUSE Tumbleweed, I'd rather have openSUSE fix it before it becomes a permanent fixture.

**Comment by Conan-Kudo** [comment]
@rkitover What Tumbleweed is doing is wrong because libtirpc on GNU systems is supposed to replace the Sun RPC headers that glibc used to ship. Putting it in a subfolder is only needed when it is not the replacement implementation yet. That's what's wrong with this.

**Comment by rkitover** [comment]
It looks like the opensuse maintainer for libtirpc may not be available right now:

https://bugzilla.opensuse.org/show_bug.cgi?id=1113116

I'll follow up on this on the factory list.

**Comment by rkitover** [comment]
Opened:

https://bugzilla.opensuse.org/show_bug.cgi?id=1122665

I also posted a message to factory, it should show up here at some point, not showing up right now for some reason:

https://lists.opensuse.org/opensuse-factory/2019-01/threads2.html

Hopefully someone will fix it quickly, if they ask me to send a PR in OBS for the package I could probably figure something out, but I'll wait before doing that because the current maintainer looks to be busy with other things.

**Comment by krop** [comment]
> Actually, it looks like the problem is subtler than that. SUSE's packaging is mutating the paths, based on what I see in [the libtirpc spec](https://build.opensuse.org/package/view_file/Base:System/libtirpc/libtirpc.spec?expand=1).
> 

I have to disagree. There's a bogus #include in ./lib/libspl/include/rpc/xdr.h:
`#include <tirpc/rpc/xdr.h>`

changing the installation path and the pkgconfig file on openSUSE would just work around this issue.
The Cflags would then be -I/usr/include/tirpc.

Suggestion: rename lib/libspl/include/rpc/xdr.h and fix the #include.

**Comment by rkitover** [comment]
The opensuse developers disagree:

https://lists.opensuse.org/opensuse-factory/2019-01/msg00348.html

To quote:

```
As one of the major driver of the sunrpc to tirpc move, even upstream,
I can only say: Closed as invalid. We will not break hundred of correct
implemented packages to workaround one broken one.

tirpc was always developed in mind as drop in replacement for sunrpc,
and we spend a lot of efford in making it as easy and simple as possible.
This includes, that you can install libtirpc in the same places as
sunrpc and 99% of the sunrpc code will compile with any change needed.
Your suggestion would break all this packages immeaditly.

Thorsten
```

**Comment by rkitover** [comment]
Looking on gentoo, libspl standalone does not come with `rpc/xdr.h`.

I could rename this header in my PR, this would make everything less hacky too, if you would like. What would be a good name?

**Comment by rkitover** [comment]
@Conan-Kudo so the opensuse developers refuse to do this (see above.) I don't know enough about these things to judge whether they are right or wrong, but I would still like to have some kind of non-manual workaround for tumbleweed users.

Would you be ok with this? And how would you like to do it?

Renaming the problematic libspl header seems like a reasonable thing to do as well, and make whatever autoconf workaround less hacky.

**Comment by Conan-Kudo** [comment]
The openSUSE developers want us to rename our file and use pkgconfig to detect libtirpc. In general, I don't see a problem with that approach.

However, this approach in this PR is not a good idea, and we should just go ahead and fix the header so that it doesn't get masked anymore.

**Comment by rkitover** [comment]
@Conan-Kudo can you please take a look at my last commit, I made things clean and I think according to everyone's intentions.

**Comment by Conan-Kudo** [comment]
@rkitover @behlendorf Is there a reason we can't rename the header in the SPL so it doesn't mask the real one?

**Comment by rkitover** [comment]
I think one reason to not rename the header is that the point of libspl is to mask system headers, so for example it has headers like `unistd.h` and `string.h`.

In this case `rpc/xdr.h` would be another system header, whether or not it comes from libtirpc. And I am using the same mechanism libspl uses for the other system headers it masks, that is `#include_next`.

**Comment by rkitover** [comment]
Can anyone else please take a look at this, I'm pretty sure it can be merged as-is.

**Comment by tonyhutter** [comment]
@rkitover looks like it's hitting some build errors.
http://build.zfsonlinux.org/builders/Ubuntu%2016.04%20i386%20%28BUILD%29/builds/9003/steps/shell_3/logs/configure

**Comment by rkitover** [comment]
@tonyhutter I saw that, but I very much doubt this has anything to do with my changes. `AC_MSG_FAILURE` is a standard autoconf macro that has been available for a long time.

**Comment by rkitover** [comment]
Also that macro was already in use in other autoconf scripts before my changes.

**Comment by behlendorf** [comment]
@rkitover it looks like the build failures are actually caused by the use of `PKG_CHECK_MODULES`.  This macro is provided by the `pkg.m4` file which is part of the `pkg-config` package.  This is not installed by default in all the build environments.  There appears to be quite a spirited discussion going back years regarding the use of `PKG_CHECK_MODULES`.

https://stackoverflow.com/questions/10220946/pkg-check-modules-considered-harmful

Rather than introduce a new dependency can we handle this another way.

**Comment by rkitover** [comment]
@behlendorf 

Can we just bundle `pkg.m4` then?

I can also just call `pkg-config` manually, that is not hard to do.

I do not agree with this SO thread. I do a lot of build related code, and using `pkg-config` is absolutely essential, on all platforms. Also that thread is 6 years old. You can avoid using it now, but you will run into this situation again. And for a library with complex `CFLAGS`/`LDFLAGS`, good luck figuring them out manually.

If you really don't want to use `pkg-config`, I can check the two previously mentioned locations and we can add `--with-tirpc=`, `--with-tirpc-includes=` and `--with-tirpc-lib`options, that is another way this is sometimes handled. It's possible in this case because we only need the `-I` `CFLAGS` option.

**Comment by rkitover** [comment]
@behlendorf @Conan-Kudo @tonyhutter 

I realized my logic was wrong wrt. checking both libc and libtirpc, this is fixed in my last commit.

I would like to finally finish this to make progress on my opensuse-on-zfs project, I'm fine with any solution for finding the library, I listed some options above.

The header definitely needed to be fixed and I did that, to bring it in line with the rest of libspl.

Please let me know how to proceed.

**Comment by rkitover** [comment]
@behlendorf @Conan-Kudo @tonyhutter

I've made some improvements for the build dependency issues:

- everything built in the CI now

- I'm including pkg.m4 in `config/` now, the license in the file states:

```
As a special exception to the GNU General Public License, if you
distribute this file as part of a program that contains a
configuration script generated by Autoconf, you may include it under
the same distribution terms that you use for the rest of that
program.
```

- `pkg.m4` does not need to be updated more than every few years

- this allows the use of `PKG_CHECK_MODULES` regardless of whether `pkg-config` is installed

- the autoconf code is simple, on `pkg-config` failure I check the standard locations and set the same variables, a similar strategy can be used for other easy to link libs

- the variables are marked precious and are described in `./configure --help` automatically by `pkg.m4`, and can be used to configure the library with or without `pkg-config`, on the `./configure` command line or in the environment, for example:

```
./configure LIBTIRPC_CFLAGS="-I/usr/include/tirpc" LIBTIRPC_LIBS="-ltirpc" --with-tirpc
```

So with this method you get: pkg-config support without it being a dependency, user-overrides and checking of standard locations with fairly simple autoconf code.

What do you think?

**Comment by rkitover** [comment]
@behlendorf If you need help setting up a docker image or some such I can probably help, I'm somewhat familiar with the process.

**Comment by rkitover** [comment]
@behlendorf can you please look at my last commit. I did some refactoring.

I made this hopefully nice `FIND_SYSTEM_LIBRARY()` macro which automates all of this. I am by no means an `autoconf`/`m4` expert, but this seems to work. Also the macro does a header and library sanity check.

Also my previous iteration was wrong, I was not quoting the variables, I tested this without `pkg-config` installed as well and it seems to work.

File: config/user-libtirpc.m4 [hunk 1]
```diff
@@ -4,27 +4,27 @@ dnl #
...
-		[if test "x$with_tirpc" != xcheck; then
-		    AC_MSG_FAILURE(
-			[--with-tirpc was given, but test for tirpc failed])
-		 fi
-		AC_SEARCH_LIBS([xdrmem_create], [tirpc], [], [
-		    AC_MSG_FAILURE([xdrmem_create() requires tirpc or libc])])
-		])],
-	    [AC_SEARCH_LIBS([xdrmem_create], [tirpc], [], [
-		AC_MSG_FAILURE([xdrmem_create() requires libc])])
+	if test "x$with_tirpc" != "xyes"; then
```
Comment (behlendorf) [inline config/user-libtirpc.m4]: Unless there's a good reason not too, we should continue to use `AS_IF` throughout.  This isn't actually cosmetic, using `AS_IF` ensures that any macros required by the conditional are expanded before the conditional.

**Comment by rkitover** [comment]
@behlendorf I've rewritten all `if` statements with `AS_IF()` and all `for` statements with `m4_foreach()`.

For the alpine test, you want to use alpine because it uses musl? Just build for whatever kernel?

I will describe the tests and the output here.

**Comment by rkitover** [comment]
@behlendorf 

Ok, going to describe testing this on alpine:

## case 1: alpine with dev tools, pkgconf, NO libtirpc-dev, linux-5.0-rc7 linked to /usr/src/linux

```
checking for library containing xdrmem_create... no
checking for libtirpc... no
configure: WARNING: cannot find libtirpc via pkg-config or in the standard locations
configure: error: in `/home/rkitover/src/zfs':
configure: error: neither libc sunrpc support nor libtirpc is available, try installing libtirpc-devel
```

## case 2: alpine with dev tools, pkgconf, libtirpc-dev, linux-5.0-rc7 linked to /usr/src/linux

```
checking for library containing xdrmem_create... no
checking for libtirpc... yes
checking rpc/xdr.h usability... yes
checking rpc/xdr.h presence... yes
checking for rpc/xdr.h... yes
checking for xdrmem_create in -ltirpc... yes
```

Build dies with:

```
/usr/lib/gcc/x86_64-alpine-linux-musl/8.2.0/../../../../x86_64-alpine-linux-musl/bin/ld: ../../../../lib/libnvpair/.libs/libnvpair.so: undefined reference to `libintl_dgettext'
```

gettext-dev is installed.

## case 3: alpine with dev tools, NO pkgconf, NO libtirpc-dev, linux-5.0-rc7 linked to /usr/src/linux

```
checking for library containing xdrmem_create... no
checking for libtirpc... no
configure: WARNING: cannot find libtirpc via pkg-config or in the standard locations
configure: error: in `/home/rkitover/src/zfs':
configure: error: neither libc sunrpc support nor libtirpc is available, try installing libtirpc-devel
```

## case 4: alpine with dev tools, NO pkgconf, libtirpc-dev, linux-5.0-rc7 linked to /usr/src/linux

```
checking for libtirpc... no
checking rpc/xdr.h usability... yes
checking rpc/xdr.h presence... yes
checking for rpc/xdr.h... yes
checking for xdrmem_create in -ltirpc... yes
```

Build dies in the same place as case 2.

## SUMMARY

Everything works as expected, need a fix for gettext on alpine (completely unrelated) I can do this in this or a different PR.

**Comment by rkitover** [comment]
@behlendorf so I almost built it in alpine with:

```
./configure LIBS=-lintl
```

but now I'm getting a different error on the kernel source that may be related to 5.0-rc7:

```
  CC [M]  /home/rkitover/src/zfs/module/spl/spl-condvar.o
/home/rkitover/src/zfs/module/spl/spl-condvar.c: In function 'spl_io_schedule_timeout':
/home/rkitover/src/zfs/module/spl/spl-condvar.c:169:2: error: implicit declaration of function 'init_timer'; did you mean 'init_timers'? [-Werror=implicit-function-declaration]
  init_timer(&timer);
  ^~~~~~~~~~
  init_timers
/home/rkitover/src/zfs/module/spl/spl-condvar.c:170:2: error: implicit declaration of function 'setup_timer'; did you mean 'del_timer'? [-Werror=implicit-function-declaration]
  setup_timer(&timer, __cv_wakeup, (unsigned long)current);
  ^~~~~~~~~~~
  del_timer
cc1: some warnings being treated as errors
```

I'm going to try a 4.20 kernel instead.

**Comment by behlendorf** [comment]
I'm mainly interesting in Alpine because of musl. There kernel issues are unrelated and easy enough to test on other distributions.

OK, thanks for doing the manual testing.  Then it sounds like things are working as expected, expect for that one unrelated issue.  Having the checks all jumbled up isn't ideal, but there's only so much you can do with autoconf.  This looks good!

**Comment by rkitover** [comment]
@behlendorf awesome! Thank you for seeing this through.

BTW I did manage to build on alpine after I copied 4.20.11 headers to `/usr/include`, I have no idea where the ones there came from and `apk` didn't know either.

Should I do something for libintl/gettext detection?

I think I can clean up the output of the macro using these things:

https://www.gnu.org/software/autoconf/manual/autoconf-2.60/html_node/File-Descriptor-Macros.html#File-Descriptor-Macros

but I can do that in a subsequent PR.

**Comment by behlendorf** [comment]
> Should I do something for libintl/gettext detection?

It would be great if you could tackle this a new PR.

> I think I can clean up the output of the macro using these things:

Yeah, let's see how horrible it really is and if it's worth making it prettier.

@Conan-Kudo would you mind re-reviewing this PR.

**Pull request merged**